polynote-spark/src/main/scala/polynote/kernel/remote/RemoteSparkKernel.scala

```diff
@@ -0,0 +1,337 @@
+    case Some(deferred) =>
+      Queue.unbounded[IO, Option[Result]] >>= deferred.complete
+  }
+
+  private def withQueue(reqId: Int, fn: Queue[IO, Option[Result]] => IO[Unit]) = Option(streamRequests.get(reqId)) match {
+    case None => IO(logger.error(s"Received stream start for $reqId, but there is no corresponding stream request outstanding"))
+    case Some(deferred) => deferred.get >>= fn
+  }
+
+  private def transformResult(result: Result): Result = result match {
+    case rv @ ResultValue(_, _, reprs, _, _, _) =>
+      // the ResultValue will be missing its JVM value and its type. Also, any handle-based reprs won't refer to handles
+      // within this JVM. We shouldn't need the value or type outside of the actual kernel, but we do have to fix
+      // the repr handles.
+      rv.copy(reprs = reprs.map {
+        case s @ StreamingDataRepr(_, _, _) => StreamingDataRepr.fromHandle(transformRepr(s))
+        case l @ LazyDataRepr(_, _) => LazyDataRepr.fromHandle(transformRepr(l))
+        case u @ UpdatingDataRepr(remoteHandle, dataType) =>
+          val local = UpdatingDataRepr(dataType)
+          updatingHandleMapping.put(remoteHandle, local.handle)
+          local
+        case repr => repr
+      })
+    case _ => result
+  }
+
+  private def consumeResponses(responses: Stream[IO, RemoteResponse]) = responses.map {
+    case Announce(remoteAddress) =>
+      Stream.eval(getNotebook().flatMap(notebook => request1[UnitResponse](InitialNotebook(_, notebook)).as(()))) ++
+        Stream.eval(remoteAddr.complete(remoteAddress).handleErrorWith(_ => IO.unit))
+    case KernelStatusResponse(update) => Stream.eval(statusUpdates.publish1(update))
+    case StreamStarted(reqId) => Stream.eval(handleStreamStart(reqId))
+    case StreamEnded(reqId) => Stream.eval(withQueue(reqId, _.enqueue1(None)))
+    case ResultStreamElements(reqId, elements) => Stream.eval(withQueue(reqId, q => Stream.emits(elements).map(Some(_)).through(q.enqueue).compile.drain))
+    case rep: RequestResponse => Stream.eval(handleOneResponse(rep))
+  }.parJoinUnbounded
+
+  private val initialized = Deferred.unsafe[IO, Either[Throwable, Unit]]
+  private val initializer = Semaphore[IO](1).unsafeRunSync()
+
+  val init: IO[Unit] = initializer.tryAcquire.flatMap {
+    case true => {
+      for {
+        mainFiber    <- consumeResponses(transport.responses).compile.drain.start
+        _            <- this.mainFiber.complete(mainFiber)
+        _            <- IO(logger.info("Started processing responses"))
+        _            <- transport.connected
+        _            <- remoteAddr.get
+      } yield ()
+    }.attempt.flatMap(initialized.complete)
+
+    case false => initialized.get.flatMap(IO.fromEither)
+  }
+
+  def shutdown(): IO[Unit] =
+    request1[UnitResponse](Shutdown(_)) *> shutdownSignal.complete *> transport.close() *> (mainFiber.get >>= (_.cancel))
```

Comment:
Is it possible that there could be a race condition where `shutdown()` is called before the `mainFiber` has been completed? In that case, might `mainFiber.get` hang? Does `Deferred` have an equivalent to `Future.isCompleted`?

Reply:
Well... if `init` is started before `shutdown` then I don't think it will deadlock, assuming `compile.drain.start` can't fail and `this.mainFiber.complete` can't fail.

`Deferred` doesn't have a way to check whether it's completed (or a nonblocking `tryGet` kinda thing). That would have often been useful, though... maybe we should implement a construct like that.

Reply:
I removed `mainFiber` since it wasn't really necessary. Instead it interrupts the stream on close. Also added `Memoize`, which I guess isn't referentially transparent (it's been discussed an rejected in cats-effect) but is useful anyway so whatever

Reply:
Actually, looks like there's `Deferred.tryable` which lets you `tryGet`. Oops.